.github/workflows/app-tests.yaml

```diff
@@ -73,4 +73,16 @@ jobs:
         - name: Run MyPy
           run: python3 -m mypy .
         - name: Run Pytest
-          run: python3 -m pytest
+          run: python3 -m pytest -s -vv --cov --cov-fail-under=85
+        - name: Run E2E tests with Playwright
+          id: e2e
+          if: runner.os != 'Windows'
+          run: |
```

Comment:
Is there a reason for not running the tests on windows?

Reply:
Never mind, It's not working for windows + it takes ~3 minutes to download the chromium browser.
https://github.com/john0isaac/rag-postgres-openai-python/actions/runs/10079481453/job/27866912673

Reply:
In azure-search-openai-demo, the Windows image has a problem with tiktoken download, which is why we often skip the unit tests. I don't remember if we explicitly tested the playwright ones on windows for that CI. But looks like you ran into issues for it. I think testing it just on Linux is fine.
We might also want to do a separate CI for playwright tests, come to think of it, as I don't think they should be different across the Python versions. What do you think?

Reply:
IMHO, I think it's not worth separating it as it will require the same setup as this workflow (without the different python versions) so we might as well do it here.
I also think it's a good thing to test on different browsers specifically (Mac and Linux) since things don't look the same on different OSs.

I will try to get it working on windows.

Reply:
After installing the browser Playwright tries to access it from an undefined folder for some reason:
```
playwright._impl._errors.Error: BrowserType.launch: ENOENT: no such file or directory, mkdtemp 'undefined\temp\playwright-artifacts-XXXXXXXXXXXX'
```

I found the error on playwright repo but the maintainers seem to close most of the issues without actually offering a valid resolution to them.

Reply:
Can you open an issue? I tried setting the path multiple times but it never works for some reason playwright tries to access it from `undefined` folder.

```yaml

name: App Tests

on:
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]
  workflow_dispatch:

permissions:
  contents: read

jobs:
  test-package:
    name: Test ${{ matrix.os }} Python ${{ matrix.python_version }}
    runs-on: ${{ matrix.os }}
    strategy:
      fail-fast: false
      matrix:
        os: ["windows-latest"]
        python_version: ["3.10", "3.11", "3.12"]
    steps:
        - uses: actions/checkout@v4
        - name: Setup python
          uses: actions/setup-python@v5
          with:
            python-version: ${{ matrix.python_version }}
            architecture: x64
        - name: Install dependencies
          run: |
            python -m pip install -r requirements-dev.txt
        - name: Install app as editable app
          run: |
            python -m pip install -e src/backend
        - name: Setup node
          uses: actions/setup-node@v4
          with:
            node-version: 18
        - name: Build frontend
          run: |
            cd ./src/frontend
            npm install
            npm run build
        - name: Run E2E tests on Windows with Playwright
          run: |
            set PLAYWRIGHT_BROWSERS_PATH=%USERPROFILE%\AppData\Local\ms-playwright
            playwright install chromium --with-deps
            python3 -m pytest tests/e2e.py --tracing=retain-on-failure
```

This is the traceback:

```bash
  set PLAYWRIGHT_BROWSERS_PATH=%USERPROFILE%\AppData\Local\ms-playwright
  playwright install chromium --with-deps
  python3 -m pytest tests/e2e.py --tracing=retain-on-failure
  shell: C:\Program Files\PowerShell\7\pwsh.EXE -command ". '{0}'"
  env:
    pythonLocation: C:\hostedtoolcache\windows\Python\3.10.11\x64
    PKG_CONFIG_PATH: C:\hostedtoolcache\windows\Python\3.10.11\x64/lib/pkgconfig
    Python_ROOT_DIR: C:\hostedtoolcache\windows\Python\3.10.11\x64
    Python2_ROOT_DIR: C:\hostedtoolcache\windows\Python\3.10.11\x64
    Python3_ROOT_DIR: C:\hostedtoolcache\windows\Python\3.10.11\x64

Success Restart Needed Exit Code      Feature Result                               
------- -------------- ---------      --------------                               
True    No             Success        {Media Foundation}                           


Downloading Chromium 127.0.6533.17 (playwright build v1124) from https://playwright.azureedge.net/builds/chromium/1124/chromium-win64.zip
|                                                                                |   0% of 134.8 MiB
|■■■■■■■■                                                                        |  10% of 134.8 MiB
|■■■■■■■■■■■■■■■■                                                                |  20% of 134.8 MiB
|■■■■■■■■■■■■■■■■■■■■■■■■                                                        |  30% of 134.8 MiB
|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■                                                |  40% of 134.8 MiB
|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■                                        |  50% of 134.8 MiB
|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■                                |  60% of 134.8 MiB
|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■                        |  70% of 134.8 MiB
|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■                |  80% of 134.8 MiB
|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■        |  90% of 134.8 MiB
|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■| 100% of 134.8 MiB
Chromium 127.0.6533.17 (playwright build v1124) downloaded to C:\Users\runneradmin\AppData\Local\ms-playwright\chromium-1124
Downloading FFMPEG playwright build v1009 from https://playwright.azureedge.net/builds/ffmpeg/1009/ffmpeg-win64.zip
|                                                                                |   1% of 1.4 MiB
|■■■■■■■■                                                                        |  10% of 1.4 MiB
|■■■■■■■■■■■■■■■■                                                                |  20% of 1.4 MiB
|■■■■■■■■■■■■■■■■■■■■■■■■                                                        |  30% of 1.4 MiB
|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■                                                |  40% of 1.4 MiB
|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■                                        |  50% of 1.4 MiB
|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■                                |  60% of 1.4 MiB
|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■                        |  70% of 1.4 MiB
|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■                |  80% of 1.4 MiB
|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■        |  90% of 1.4 MiB
|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■| 100% of 1.4 MiB
FFMPEG playwright build v1009 downloaded to C:\Users\runneradmin\AppData\Local\ms-playwright\ffmpeg-1009
============================= test session starts =============================
platform win32 -- Python 3.10.11, pytest-8.3.2, pluggy-1.5.0
rootdir: D:\a\rag-postgres-openai-python\rag-postgres-openai-python
configfile: pyproject.toml
plugins: anyio-4.4.0, asyncio-0.23.8, base-url-2.1.0, cov-5.0.0, playwright-0.5.1, snapshot-0.9.0
asyncio: mode=strict
collected 4 items

tests\e2e.py EEEE                                                        [100%]

=================================== ERRORS ====================================
____________________ ERROR at setup of test_home[chromium] ____________________

launch_browser = <function launch_browser.<locals>.launch at 0x0000021A4C7FC9D0>

    @pytest.fixture(scope="session")
    def browser(launch_browser: Callable[[], Browser]) -> Generator[Browser, None, None]:
>       browser = launch_browser()

C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\pytest_playwright\pytest_playwright.py:266: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\pytest_playwright\pytest_playwright.py:258: in launch
    browser = browser_type.launch(**launch_options)
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\playwright\sync_api\_generated.py:14155: in launch
    self._sync(
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\playwright\_impl\_browser_type.py:94: in launch
    Browser, from_channel(await self._channel.send("launch", params))
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\playwright\_impl\_connection.py:59: in send
    return await self._connection.wrap_api_call(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <playwright._impl._connection.Connection object at 0x0000021A4C5A5780>
cb = <function Channel.send.<locals>.<lambda> at 0x0000021A4C5CAC20>
is_internal = False

    async def wrap_api_call(
        self, cb: Callable[[], Any], is_internal: bool = False
    ) -> Any:
        if self._api_zone.get():
            return await cb()
        task = asyncio.current_task(self._loop)
        st: List[inspect.FrameInfo] = getattr(task, "__pw_stack__", inspect.stack())
        parsed_st = _extract_stack_trace_information_from_stack(st, is_internal)
        self._api_zone.set(parsed_st)
        try:
            return await cb()
        except Exception as error:
>           raise rewrite_error(error, f"{parsed_st['apiName']}: {error}") from None
E           playwright._impl._errors.Error: BrowserType.launch: ENOENT: no such file or directory, mkdtemp 'undefined\temp\playwright-artifacts-XXXXXXXXXXXX'

C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\playwright\_impl\_connection.py:514: Error
____________________ ERROR at setup of test_chat[chromium] ____________________

launch_browser = <function launch_browser.<locals>.launch at 0x0000021A4C7FC9D0>

    @pytest.fixture(scope="session")
    def browser(launch_browser: Callable[[], Browser]) -> Generator[Browser, None, None]:
>       browser = launch_browser()

C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\pytest_playwright\pytest_playwright.py:266: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\pytest_playwright\pytest_playwright.py:258: in launch
    browser = browser_type.launch(**launch_options)
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\playwright\sync_api\_generated.py:14155: in launch
    self._sync(
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\playwright\_impl\_browser_type.py:94: in launch
    Browser, from_channel(await self._channel.send("launch", params))
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\playwright\_impl\_connection.py:59: in send
    return await self._connection.wrap_api_call(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <playwright._impl._connection.Connection object at 0x0000021A4C5A5780>
cb = <function Channel.send.<locals>.<lambda> at 0x0000021A4C5CAC20>
is_internal = False

    async def wrap_api_call(
        self, cb: Callable[[], Any], is_internal: bool = False
    ) -> Any:
        if self._api_zone.get():
            return await cb()
        task = asyncio.current_task(self._loop)
        st: List[inspect.FrameInfo] = getattr(task, "__pw_stack__", inspect.stack())
        parsed_st = _extract_stack_trace_information_from_stack(st, is_internal)
        self._api_zone.set(parsed_st)
        try:
            return await cb()
        except Exception as error:
>           raise rewrite_error(error, f"{parsed_st['apiName']}: {error}") from None
E           playwright._impl._errors.Error: BrowserType.launch: ENOENT: no such file or directory, mkdtemp 'undefined\temp\playwright-artifacts-XXXXXXXXXXXX'

C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\playwright\_impl\_connection.py:514: Error
_____________ ERROR at setup of test_chat_customization[chromium] _____________

launch_browser = <function launch_browser.<locals>.launch at 0x0000021A4C7FC9D0>

    @pytest.fixture(scope="session")
    def browser(launch_browser: Callable[[], Browser]) -> Generator[Browser, None, None]:
>       browser = launch_browser()

C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\pytest_playwright\pytest_playwright.py:266: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\pytest_playwright\pytest_playwright.py:258: in launch
    browser = browser_type.launch(**launch_options)
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\playwright\sync_api\_generated.py:14155: in launch
    self._sync(
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\playwright\_impl\_browser_type.py:94: in launch
    Browser, from_channel(await self._channel.send("launch", params))
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\playwright\_impl\_connection.py:59: in send
    return await self._connection.wrap_api_call(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <playwright._impl._connection.Connection object at 0x0000021A4C5A5780>
cb = <function Channel.send.<locals>.<lambda> at 0x0000021A4C5CAC20>
is_internal = False

    async def wrap_api_call(
        self, cb: Callable[[], Any], is_internal: bool = False
    ) -> Any:
        if self._api_zone.get():
            return await cb()
        task = asyncio.current_task(self._loop)
        st: List[inspect.FrameInfo] = getattr(task, "__pw_stack__", inspect.stack())
        parsed_st = _extract_stack_trace_information_from_stack(st, is_internal)
        self._api_zone.set(parsed_st)
        try:
            return await cb()
        except Exception as error:
>           raise rewrite_error(error, f"{parsed_st['apiName']}: {error}") from None
E           playwright._impl._errors.Error: BrowserType.launch: ENOENT: no such file or directory, mkdtemp 'undefined\temp\playwright-artifacts-XXXXXXXXXXXX'

C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\playwright\_impl\_connection.py:514: Error
_____________ ERROR at setup of test_chat_nonstreaming[chromium] ______________

launch_browser = <function launch_browser.<locals>.launch at 0x0000021A4C7FC9D0>

    @pytest.fixture(scope="session")
    def browser(launch_browser: Callable[[], Browser]) -> Generator[Browser, None, None]:
>       browser = launch_browser()

C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\pytest_playwright\pytest_playwright.py:266: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\pytest_playwright\pytest_playwright.py:258: in launch
    browser = browser_type.launch(**launch_options)
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\playwright\sync_api\_generated.py:14155: in launch
    self._sync(
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\playwright\_impl\_browser_type.py:94: in launch
    Browser, from_channel(await self._channel.send("launch", params))
C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\playwright\_impl\_connection.py:59: in send
    return await self._connection.wrap_api_call(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <playwright._impl._connection.Connection object at 0x0000021A4C5A5780>
cb = <function Channel.send.<locals>.<lambda> at 0x0000021A4C5CAC20>
is_internal = False

    async def wrap_api_call(
        self, cb: Callable[[], Any], is_internal: bool = False
    ) -> Any:
        if self._api_zone.get():
            return await cb()
        task = asyncio.current_task(self._loop)
        st: List[inspect.FrameInfo] = getattr(task, "__pw_stack__", inspect.stack())
        parsed_st = _extract_stack_trace_information_from_stack(st, is_internal)
        self._api_zone.set(parsed_st)
        try:
            return await cb()
        except Exception as error:
>           raise rewrite_error(error, f"{parsed_st['apiName']}: {error}") from None
E           playwright._impl._errors.Error: BrowserType.launch: ENOENT: no such file or directory, mkdtemp 'undefined\temp\playwright-artifacts-XXXXXXXXXXXX'

C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\site-packages\playwright\_impl\_connection.py:514: Error
=========================== short test summary info ===========================
ERROR tests/e2e.py::test_home[chromium] - playwright._impl._errors.Error: BrowserType.launch: ENOENT: no such file or directory, mkdtemp 'undefined\temp\playwright-artifacts-XXXXXXXXXXXX'
ERROR tests/e2e.py::test_chat[chromium] - playwright._impl._errors.Error: BrowserType.launch: ENOENT: no such file or directory, mkdtemp 'undefined\temp\playwright-artifacts-XXXXXXXXXXXX'
ERROR tests/e2e.py::test_chat_customization[chromium] - playwright._impl._errors.Error: BrowserType.launch: ENOENT: no such file or directory, mkdtemp 'undefined\temp\playwright-artifacts-XXXXXXXXXXXX'
ERROR tests/e2e.py::test_chat_nonstreaming[chromium] - playwright._impl._errors.Error: BrowserType.launch: ENOENT: no such file or directory, mkdtemp 'undefined\temp\playwright-artifacts-XXXXXXXXXXXX'
============================== 4 errors in 1.44s ==============================
```

Reply:
resolved here #68